src/Serilog.Sinks.Map/Sinks/Map/MappedSink`1.cs

```diff
@@ -0,0 +1,116 @@
+﻿// Serilog.Sinks.Seq Copyright 2017 Serilog Contributors
+// 
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// 
+//     http://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+using System.Collections.Generic;
+using Serilog.Configuration;
+using Serilog.Core;
+using Serilog.Events;
+
+namespace Serilog.Sinks.Map
+{
+    class MappedSink<TKey> : ILogEventSink, IDisposable
+    {
+        readonly Func<LogEvent, TKey> _keySelector;
+        readonly Action<TKey, LoggerSinkConfiguration> _configure;
+        readonly int? _sinkMapCountLimit;
+        readonly object _sync = new object();
+        readonly Dictionary<TKey, Logger> _sinkMap = new Dictionary<TKey, Logger>();
+        bool _disposed;
+
+        public MappedSink(Func<LogEvent, TKey> keySelector,
+                          Action<TKey, LoggerSinkConfiguration> configure,
+                          int? sinkMapCountLimit)
+        {
+            _keySelector = keySelector;
+            _configure = configure;
+            _sinkMapCountLimit = sinkMapCountLimit;
+        }
+
+        public void Emit(LogEvent logEvent)
+        {
+            var key = _keySelector(logEvent);
+
```

Comment:
Did you consider of using `bool KeySelector<TKey>(LogEvent logEvent, out TKey key)` delegate for the `_keySelector`? It will allow `Emit` to be efficient for missing keys
```csharp
if (!_keySelector(logEvent, out var key)) {
     return;
}
```

`Map`'s overloads with `Func<LogEvent, TKey>` delegates along with the `defaultKey`'s can be adopted to this type of delegate, and new overloads can be added without `defaultKey`'s. Thoughts?

Reply:
Also a great point 👍